exporter/elasticsearchexporter/elasticsearch_bulk.go

```diff
@@ -226,33 +206,40 @@ type bulkIndexerStats struct {
 }
 
 type bulkIndexerPool struct {
-	items chan esBulkIndexerItem
-	wg    sync.WaitGroup
-	stats bulkIndexerStats
+	closeCh   chan struct{}
+	stats     bulkIndexerStats
+	available chan *worker
 }
 
-// Add adds an item to the bulk indexer pool.
-//
-// Adding an item after a call to Close() will panic.
-func (p *bulkIndexerPool) Add(ctx context.Context, index string, document io.WriterTo) error {
-	item := esBulkIndexerItem{
-		Index: index,
-		Body:  document,
-	}
+func (p *bulkIndexerPool) AddBatchAndFlush(ctx context.Context, batch []esBulkIndexerItem) error {
```

Comment:
Should we be getting rid of the `num_workers` config setting, and changing the `Export` calls to directly/synchronously send a bulk request to ES? Then all batching and queuing would only occur in `exporterhelper`. Or is that coming later?

Reply:
I agree that `num_workers` is awkward and is likely an unnecessary knob. It makes sense to remove it, but are we worried about the number of concurrent requests to ES and reuse of underlying connections? (I'm thinking estransport conn pool, but I'm not familiar with it). Theoretically, concurrency will be bound by `sending_queue.num_consumers` and likely much lower than that due to batching.

Reply:
Previously we would have a maximum of `num_workers` concurrent bulk requests, right?

Now if `sending_queue` and `batcher` were enabled and we didn't have additional buffering in the exporter, the max concurrency would be `num_consumers`. That's because there are `num_consumers` queue consumers, and `num_consumers` max concurrency set in the batch sender.

On the one hand I think 100 concurrent bulk requests is likely too high. On the other hand, I think lowering it means that the size/time thresholds may not be reached as discussed above. Doesn't seem like a great (or fundamentally necessary?) trade off.

We could introduce a semaphore to control the number of concurrent bulk requests, but I'd rather not. That should ideally just be controlled by the batch sender's max concurrency, and a queue consumer should be able to keep filling up batches independently of the batch sender's max concurrency.

Reply:
I tried to implement this in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32632/commits/18e9ce412912e9c8fc0d75b18c534af2afcd2591

2 things I've noticed:
- After removing `num_workers`, if sending_queue isn't configured, there will be unlimited concurrency. vs Before removing `num_workers`, `send` will block and wait for available worker, effectively giving back-pressure.
- Minor point: after removing num_workers, we are also removing the pooling, so it will likely hurt performance. Might need to add back a sync.Pool.

Reply:
> After removing num_workers, if sending_queue isn't configured, there will be unlimited concurrency. vs Before removing num_workers, send will block and wait for available worker, effectively giving back-pressure.

Is that because of the [batchSender.concurrencyLimit behaviour](https://github.com/open-telemetry/opentelemetry-collector/blob/0d6e6bf8643cf7b88b37aa84aa62b2b3fbd44e51/exporter/exporterhelper/batch_sender.go#L30-L33)?

```
// concurrencyLimit is the maximum number of goroutines that can be blocked by the batcher.
// If this number is reached and all the goroutines are busy, the batch will be sent right away.
// Populated from the number of queue consumers if queue is enabled.
concurrencyLimit uint64
```

I'm not sure if/why this behaviour would be desirable. Seems to me there should at least be an option for the batch sender to block when the concurrent limit is reached.

> Minor point: after removing num_workers, we are also removing the pooling, so it will likely hurt performance. Might need to add back a sync.Pool.

++ I think having a sync.Pool of BatchIndexers makes sense.

Reply:
Perhaps we can just use `num_workers` to create a [weighted semaphore](https://pkg.go.dev/golang.org/x/sync/semaphore#NewWeighted) for now? And continue initialising that to NumCPUs? Then we can figure out whether this batch sender behaviour should change, and how to change it.

Reply:
> Is that because of the [batchSender.concurrencyLimit behaviour](https://github.com/open-telemetry/opentelemetry-collector/blob/0d6e6bf8643cf7b88b37aa84aa62b2b3fbd44e51/exporter/exporterhelper/batch_sender.go#L30-L33)?

It is related. There are a few things at play here:
- `sending_queue.num_consumers`: the number of concurrent goroutines handling the lifecycle of a queue entry, as mentioned in another comment. This is only relevant when sending_queue is enabled.
- batch_sender `concurrencyLimit`:
  - When sending_queue is enabled, it is set to num_consumers (e.g. 10), reason being if there can only be e.g. 10 goroutines sending requests from queue to batch_sender, even if the batch does not reach min_size_items, there is no reason to wait for `batcher.flush_timeout` for the 11th goroutine to join in with more items, because there won't be another goroutine. It does not make sense to wait meaninglessly before sending out the batch.
  - This is not configurable from config, and if sending_queue is disabled, there will be no concurrency control. Batch sender will either wait for batcher.flush_timeout or batcher.min_size_items conditions to be satisfied before sending a batch, without the additional concurrency condition above.
  - This has nothing to do with the number of concurrent batches (i.e. bulk request) being sent.
- our very own `num_workers` config: controls the number of concurrent in-flight bulk request

Reply:
> Perhaps we can just use num_workers to create a [weighted semaphore](https://pkg.go.dev/golang.org/x/sync/semaphore#NewWeighted) for now? And continue initialising that to NumCPUs?

Done. This makes it easier to remove `num_workers` in the future.

> ++ I think having a sync.Pool of BatchIndexers makes sense.

Holding off premature optimization until the PR is stable. Might even defer it to another PR. I believe http connection will be reused anyway since it is handled by estransport, so we're talking about just the allocs of bulk indexer struct here.

When adding sync.Pool we'll have to double check for correctness, e.g. the stateful backoff function, whether buffers are clean, etc

Reply:
Now that `num_workers` keeps in flight bulk requests bounded, could we enable the queue by default? Any reasons left not to?

Reply:
Enabled sending queue by default. Added sync.Pool for bulk indexer.